File: apps/desktop/src/tauri/window.ts
/**
 * Window helpers.
 *
 * REDBEAM runs as a multi-window Tauri app, but every window loads the same
 * frontend bundle — it works out what it is from the query string. That means
 * the app also runs in a plain browser tab, which is how iteration stays fast.
 * Everything here degrades to browser behaviour when Tauri is absent rather
 * than throwing.
 */

import { invoke } from '@tauri-apps/api/core'

export type WindowRole = 'main' | 'context'

export interface WindowIdentity {
  /** null when no project is addressed — e.g. a bare `npm run dev` tab. */
  projectId: string | null
  role: WindowRole
  /**
   * The document a context window was opened to show, by relative path.
   *
   * Popping a tab out and landing on a different sheet than the one you popped
   * is worse than not having the feature, so the target travels on the URL
   * rather than the new window guessing. null on a main window.
   */
  documentPath: string | null
}

/** What the Rust side reports back after opening (or focusing) a window. */
export interface OpenedWindow {
  label: string
  url: string
  /** false when a window with that label already existed and was focused. */
  created: boolean
}

/**
 * True when running inside a Tauri webview.
 *
 * Detected via the internals object Tauri injects rather than by importing
 * `isTauri` from the API package, so this stays a cheap synchronous check with
 * no import-time cost in a plain browser.
 */
export function isTauri(): boolean {
  return typeof window !== 'undefined' && '__TAURI_INTERNALS__' in window
}

/**
 * Read this window's identity out of the query string.
 *
 * Defaults to `{ projectId: null, role: 'main' }`, which is what a plain
 * `http://localhost:5180/` tab gets.
 */
export function getWindowRole(search?: string): WindowIdentity {
  const raw =
    search ?? (typeof window === 'undefined' ? '' : window.location.search)
  const params = new URLSearchParams(raw)
  const projectId = params.get('projectId')
  const role = params.get('role')
  const document = params.get('document')
  return {
    projectId: projectId === null || projectId === '' ? null : projectId,
    role: role === 'context' ? 'context' : 'main',
    documentPath: document === null || document === '' ? null : document,
  }
}

/** Build the in-app URL a window with this identity should load. */
export function windowUrl(
  projectId: string,
  role: WindowRole,
  documentPath?: string,
): string {
  const params = new URLSearchParams({ projectId, role })
  if (documentPath !== undefined && documentPath !== '') {
    params.set('document', documentPath)
  }
  return `index.html?${params.toString()}`
}

function browserFallback(
  projectId: string,
  role: WindowRole,
  label: string,
  documentPath?: string,
): OpenedWindow {
  const params = new URLSearchParams({ projectId, role })
  if (documentPath !== undefined && documentPath !== '') {
    params.set('document', documentPath)
  }
  const url = `${window.location.pathname}?${params.toString()}`
  const opened = window.open(url, label)
  return { label, url, created: opened !== null }
}

/**
 * Open — or focus, if it is already open — a window showing a project.
 *
 * @param label window identity. Defaults to `project-<projectId>`, so calling
 *   twice for the same project focuses rather than duplicates.
 */
export async function openProjectWindow(
  projectId: string,
  label = `project-${projectId}`,
): Promise<OpenedWindow> {
  if (!isTauri()) return browserFallback(projectId, 'main', label)
  return invoke<OpenedWindow>('open_project_window', { projectId, label })
}

/**
 * Open — or focus — the context window that rides alongside a project.
 *
 * `documentPath` is the drawing it should show. There is one context window per
 * project, by label, so popping out a second tab re-points the existing window
 * rather than stacking a third.
 */
export async function openContextWindow(
  projectId: string,
  documentPath?: string,
): Promise<OpenedWindow> {
  if (!isTauri()) {
    return browserFallback(projectId, 'context', `context-${projectId}`, documentPath)
  }
  return invoke<OpenedWindow>('open_context_window', {
    projectId,
    document: documentPath ?? null,
  })
}

// ------------------------------------------------------- this window --

/**
 * Controls for the window this code is running in.
 *
 * The API package is imported dynamically rather than at module scope: this
 * file is also loaded by a plain browser tab and by the tests, where pulling in
 * the window API would cost an import for something that can never be called.
 * Every one of these is a no-op outside Tauri rather than a throw — a browser
 * tab has no window to minimize and that is not an error.
 */
async function thisWindow() {
  if (!isTauri()) return null
  const { getCurrentWindow } = await import('@tauri-apps/api/window')
  return getCurrentWindow()
}

/**
 * Report rather than swallow.
 *
 * These are core commands, so they go through Tauri's ACL: without
 * `core:window:allow-minimize` and friends in `capabilities/default.json`,
 * every one of them REJECTS. The buttons were wired to `void minimize()`,
 * which discards the rejection — so the frame's controls did nothing at all,
 * silently, and looked like dead pixels rather than a missing grant. A console
 * error naming the permission is the difference between a five-minute fix and
 * an afternoon.
 */
async function windowAction(name: string, run: () => Promise<void>): Promise<boolean> {
  try {
    await run()
    return true
  } catch (err) {
    console.error(
      `[window] ${name} was refused. If this is a desktop build, check that ` +
        `core:window:allow-${name.toLowerCase()} is granted in ` +
        'src-tauri/capabilities/default.json.',
      err,
    )
    return false
  }
}

export async function minimizeWindow(): Promise<boolean> {
  const w = await thisWindow()
  if (w === null) return false
  return windowAction('minimize', () => w.minimize())
}

/** Maximize, or restore if already maximized. Resolves the state afterwards. */
export async function toggleMaximizeWindow(): Promise<boolean> {
  const w = await thisWindow()
  if (w === null) return false
  await windowAction('toggle-maximize', () => w.toggleMaximize())
  return isWindowMaximized()
}

export async function closeThisWindow(): Promise<boolean> {
  const w = await thisWindow()
  if (w === null) return false
  return windowAction('close', () => w.close())
}

export async function isWindowMaximized(): Promise<boolean> {
  const w = await thisWindow()
  if (w === null) return false
  try {
    return await w.isMaximized()
  } catch {
    return false
  }
}

/**
 * Subscribe to resize, which is the only event that changes maximized state.
 *
 * Resolves a function that unsubscribes, or a no-op outside Tauri.
 */
export async function onWindowResize(fn: () => void): Promise<() => void> {
  const w = await thisWindow()
  if (w === null) return () => {}
  return w.onResized(fn)
}

/** Close a window by label. Resolves false when there was nothing to close. */
export async function closeWindow(label: string): Promise<boolean> {
  if (!isTauri()) {
    window.close()
    return true
  }
  return invoke<boolean>('close_window', { label })
}

File: apps/desktop/src-tauri/src/window.rs
//! Additional-window creation.
//!
//! REDBEAM is a multi-window app: a project window per open project, plus an
//! optional "context" window that rides alongside one. Every window loads the
//! same frontend bundle and works out what it is from the query string, so
//! there is exactly one entry point to build and ship.

use tauri::{AppHandle, Manager, WebviewUrl, WebviewWindow, WebviewWindowBuilder};

/// Describes a window that was opened (or focused, if it already existed).
#[derive(Debug, Clone, serde::Serialize)]
pub struct OpenedWindow {
    pub label: String,
    pub url: String,
    /// False when an existing window with the same label was focused instead.
    pub created: bool,
}

/// Percent-encode a value for use inside a query string.
///
/// Written out rather than pulled in as a dependency: the alphabet is tiny and
/// this keeps the crate's dependency list to Tauri plus serde.
fn encode_query_value(value: &str) -> String {
    let mut out = String::with_capacity(value.len());
    for byte in value.as_bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(*byte as char)
            }
            _ => out.push_str(&format!("%{:02X}", byte)),
        }
    }
    out
}

/// Tauri only accepts `[a-zA-Z0-9-/:_]` in window labels, so anything else in a
/// caller-supplied label becomes `_`.
fn sanitize_label(label: &str) -> String {
    let cleaned: String = label
        .chars()
        .map(|c| {
            if c.is_ascii_alphanumeric() || matches!(c, '-' | '/' | ':' | '_') {
                c
            } else {
                '_'
            }
        })
        .collect();
    if cleaned.is_empty() {
        "window".to_string()
    } else {
        cleaned
    }
}

fn app_url(project_id: &str, role: &str) -> String {
    format!(
        "index.html?projectId={}&role={}",
        encode_query_value(project_id),
        encode_query_value(role)
    )
}

/// Build a window, or focus the one that already owns this label.
///
/// Must not be called from the main thread — on Windows the builder blocks
/// waiting on the event loop it would be sitting on. The commands below are
/// `async`, which is what keeps this off the main thread.
fn open_or_focus(
    app: &AppHandle,
    label: &str,
    url: &str,
    title: &str,
) -> Result<OpenedWindow, tauri::Error> {
    if let Some(existing) = app.get_webview_window(label) {
        // Re-point it. A context window is a VIEW, and popping out a second tab
        // asks for that view to show the second document — focusing a window
        // still displaying the first one looks like the command did nothing.
        // Resolved against the window's CURRENT url so this stays correct under
        // both `tauri dev` (an http origin) and a bundled build (a custom
        // scheme). If either step fails we simply focus what is there — better
        // than failing the command outright.
        if let Ok(target) = existing.url().ok().ok_or(()).and_then(|b| b.join(url).map_err(|_| ())) {
            let _ = existing.navigate(target);
        }
        let _ = existing.unminimize();
        let _ = existing.show();
        let _ = existing.set_focus();
        return Ok(OpenedWindow {
            label: label.to_string(),
            url: url.to_string(),
            created: false,
        });
    }

    let window: WebviewWindow =
        WebviewWindowBuilder::new(app, label, WebviewUrl::App(url.into()))
            .title(title)
            .inner_size(1280.0, 820.0)
            // The same floor tauri.conf.json gives the first window, and for
            // the same reason: the shell's grid is 44 + 232 + 360 + 320 = 956
            // of irreducible columns, so a window narrower than that pushes
            // the workspace pane off its own right edge. A second project
            // window used to be built at 800 and did exactly that.
            .min_inner_size(960.0, 640.0)
            .resizable(true)
            // The app draws its own title bar; `shadow` is what keeps the
            // resize border and Windows snap on an undecorated window.
            .decorations(false)
            .shadow(true)
            // Tauri's own default arguments, restated because setting this
            // replaces them rather than adding to them. Deliberately NOT
            // `--disable-pinch`: suppressing the gesture upstream can only take
            // it away from the page that wants to handle it.
            .additional_browser_args(
                "--disable-features=msWebOOUI,msPdfOOUI,msSmartScreenProtection",
            )
            /*
             * THIS IS WHAT MAKES A TRACKPAD PINCH REACH THE PAGE.
             *
             * The name says hotkeys. On Windows, wry passes this straight to
             * `ICoreWebView2Settings5::SetIsPinchZoomEnabled` — and its default
             * is FALSE, so a Tauri window that says nothing about zoom is a
             * window where WebView2 does not recognise the pinch gesture at
             * all. Not "recognises it and scales the page": does not generate
             * it. The ctrl+wheel that Chromium synthesizes for a precision
             * touchpad pinch is never produced, and the page receives plain
             * scroll wheels indistinguishable from a two-finger swipe.
             *
             * Measured, rather than reasoned: the same page in a browser tab on
             * the same machine and trackpad produced 633 wheel events with
             * ctrl=1; in this webview, 146 events and not one of them.
             *
             * Turning it on does NOT hand zoom to the webview. The page claims
             * every ctrl+wheel in the document and every zoom key, so WebView2
             * generates the gesture and we are the ones who act on it. See the
             * zoom-hotkey effect in Workspace.tsx.
             */
            .zoom_hotkeys_enabled(true)
            .center()
            .build()?;

    // Focus is the whole point of opening a window — except under the
    // minimized switch, where taking focus is exactly what we are avoiding.
    if minimize_if_requested(&window) {
        // minimized instead
    } else {
        let _ = window.set_focus();
    }

    Ok(OpenedWindow {
        label: label.to_string(),
        url: url.to_string(),
        created: true,
    })
}

/// Open (or focus) a window showing a project.
/// Honour REDBEAM_START_MINIMIZED for windows created after startup too.
///
/// The setup hook only sees windows that exist when the app boots; a context
/// window opened later would still raise itself over whatever the user is
/// doing, which is exactly what the switch exists to prevent.
fn minimize_if_requested(window: &tauri::WebviewWindow) -> bool {
    let on = std::env::var("REDBEAM_START_MINIMIZED")
        .map(|v| {
            let v = v.trim().to_ascii_lowercase();
            !(v.is_empty() || v == "0" || v == "false" || v == "no")
        })
        .unwrap_or(false);
    if on {
        let _ = window.minimize();
    }
    on
}

#[tauri::command]
pub async fn open_project_window(
    app: AppHandle,
    project_id: String,
    label: String,
) -> Result<OpenedWindow, String> {
    let label = sanitize_label(&label);
    let url = app_url(&project_id, "main");
    let title = format!("REDBEAM — {}", project_id);
    open_or_focus(&app, &label, &url, &title).map_err(|e| e.to_string())
}

/// Open (or focus) the context window for a project.
///
/// `document` is the drawing to show, carried on the URL. Popping a tab out and
/// landing on a different sheet than the one you popped is worse than not
/// having the feature, so the target travels with the request rather than the
/// new window guessing.
///
/// One context window per project, by label — a second pop-out re-points the
/// existing one instead of stacking windows.
#[tauri::command]
pub async fn open_context_window(
    app: AppHandle,
    project_id: String,
    document: Option<String>,
) -> Result<OpenedWindow, String> {
    let label = sanitize_label(&format!("context-{}", project_id));
    let mut url = app_url(&project_id, "context");
    if let Some(doc) = document.as_deref().filter(|d| !d.is_empty()) {
        url.push_str(&format!("&document={}", encode_query_value(doc)));
    }
    let title = format!("REDBEAM Context — {}", project_id);
    open_or_focus(&app, &label, &url, &title).map_err(|e| e.to_string())
}

/// Close a window by label. Returns false if there was nothing to close.
#[tauri::command]
pub async fn close_window(app: AppHandle, label: String) -> Result<bool, String> {
    match app.get_webview_window(&sanitize_label(&label)) {
        Some(window) => {
            window.close().map_err(|e| e.to_string())?;
            Ok(true)
        }
        None => Ok(false),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn encodes_query_values() {
        assert_eq!(encode_query_value("abc-123_x.y~z"), "abc-123_x.y~z");
        assert_eq!(encode_query_value("a b&c=d"), "a%20b%26c%3Dd");
    }

    #[test]
    fn sanitizes_labels() {
        assert_eq!(sanitize_label("project-42"), "project-42");
        assert_eq!(sanitize_label("bad label!"), "bad_label_");
        assert_eq!(sanitize_label(""), "window");
    }

    #[test]
    fn builds_app_urls() {
        assert_eq!(
            app_url("p 1", "context"),
            "index.html?projectId=p%201&role=context"
        );
    }
}
